README
The Debian Package vyatta-op-dhcp-server
-------------------------------------

This package has the CLI template for the "clear dhcp" commands


 -- Marat Nepomnyashy <marat@vyatta.com> Thu 27 Dec 2007


debian/changelog
vyatta-op-dhcp-server (0.14.0+vyos2+current2) unstable; urgency=medium

  * Use ISC DHCP servers instead of the vyatta version.

 -- Daniil Baturin <daniil@baturin.org>  Wed, 27 Jan 2016 07:55:00 -0500

vyatta-op-dhcp-server (0.14.0+vyos2+current1) unstable; urgency=medium

  [ Thomas Jepp ]
  * Fix build depends.

  [ Kim Hagen ]

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 15:19:57 -0500

vyatta-op-dhcp-server (0.14.0+vyos2+lithium6) unstable; urgency=low

  * vyatta-op-dhcp-server: update dh_gencontrol with new development flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 19 Jun 2015 20:28:54 +0100
 
vyatta-op-dhcp-server (0.14.0+vyos2+lithium5) unstable; urgency=low

  [ Dominik Malis ]
  * Bug 520 - Feature request: Show DUID of DHCPv6 client request in log

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 07 Feb 2015 10:29:46 +0000
  
vyatta-op-dhcp-server (0.14.0+vyos2+lithium4) unstable; urgency=low

  [ Patrick van Staveren ]
  * vyatta-op-dhcp-server: "Malformed lease" with an abandoned DHCP lease

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 07 Feb 2015 10:29:46 +0000
 
vyatta-op-dhcp-server (0.14.0+vyos2+lithium3) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 27 Dec 2014 11:32:56 +0000
 
vyatta-op-dhcp-server (0.14.0+vyos2+lithium2) unstable; urgency=low

  * Force release

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 15 Dec 2014 18:25:33 +0000
 
vyatta-op-dhcp-server (0.14.0+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 1 Dec 2014 07:31:42 +0000
 
vyatta-op-dhcp-server (0.14.0+vyos1+helium4) unstable; urgency=low

  [Alex Harpin]
  * vyatta-op-dhcp-server: tidy up output on "show dhcp server leases"
  * vyatta-op-dhcp-server: formatting changes for style consistency
  * vyatta-op-dhcp-server: update dhcp lease parsing logic based on type
  * vyatta-op-dhcp-server: strip shared-network prefix from client-hostname

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 1 Dec 2014 07:16:58 +0000
 
vyatta-op-dhcp-server (0.14.0+vyos1+helium3) unstable; urgency=low

  * Bug #122: Move DHCPv6 lease file to /config to make it survive image
    upgrades.

 -- Daniil Baturin <daniil@baturin.org>  Wed, 15 Jan 2014 05:19:37 +0100

vyatta-op-dhcp-server (0.14.0+vyos1+helium2) unstable; urgency=low

  * Bug #121: fix lease regex so it doesn't catch anything but well-
    formed leases.
  * Bug #121: don't require all fields to be present, they may be absent

 -- Daniil Baturin <daniil@baturin.org>  Wed, 15 Jan 2014 01:47:02 +0100

vyatta-op-dhcp-server (0.14.0+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Wed, 15 Jan 2014 01:07:59 +0100

vyatta-op-dhcp-server (0.14.0+hydrogen3) unstable; urgency=low

  * Change architecture from 'any' to 'all', as package now has no
  * Update debian package maintainer address

 -- Daniil Baturin <daniil@baturin.org>  Sun, 10 Nov 2013 15:17:21 +0100

vyatta-op-dhcp-server (0.14.0+hydrogen2) unstable; urgency=low

  * Update maintainers address in autoconf rules
  * Replace XSL-based op mode commands with plain Perl
  * Update package dependencies
  * Remove now-unused sablotron-based code and related rules

 -- Daniil Baturin <daniil@baturin.org>  Sun, 10 Nov 2013 15:11:08 +0100

vyatta-op-dhcp-server (0.14.0+hydrogen1) unstable; urgency=low

  * Fix branch name

 -- Daniil Baturin <daniil@baturin.org>  Sun, 10 Nov 2013 00:00:00 +0100

vyatta-op-dhcp-server (0.13.24+daisy2) unstable; urgency=low

  * Check that "dhcp server" is configured before showing stats

 -- James Davidson <james.davidson@vyatta.com>  Fri, 24 May 2013 13:40:17 -0700

vyatta-op-dhcp-server (0.13.24+daisy1) unstable; urgency=low

  * create daisy branch

 -- John Southworth <john.southworth@vyatta.com>  Sat, 13 Oct 2012 13:31:15 -0700

vyatta-op-dhcp-server (0.13.24) unstable; urgency=low

  * new branch

 -- John Southworth <john.southworth@vyatta.com>  Fri, 12 Oct 2012 19:47:43 -0700

vyatta-op-dhcp-server (0.13.23) unstable; urgency=low

  * Move dhcpd.leases to /config so it is persistent

 -- James Davidson <james.davidson@vyatta.com>  Mon, 17 Sep 2012 19:05:04 -0700

vyatta-op-dhcp-server (0.13.22) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:26:10 -0800

vyatta-op-dhcp-server (0.13.21) unstable; urgency=low

  * Bugfix 7483: Make DHCP server op mode under a server node

 -- John Southworth <john.southworth@vyatta.com>  Tue, 28 Feb 2012 15:51:04 -0800

vyatta-op-dhcp-server (0.13.20) unstable; urgency=low

  * Fix off by one problem in DHCP server op mode commands

 -- John Southworth <john.southworth@vyatta.com>  Thu, 02 Feb 2012 08:49:02 -0800

vyatta-op-dhcp-server (0.13.19) unstable; urgency=low

  * Add ability to restart ipv4 dhcp server and relay-agent

 -- John Southworth <john.southworth@vyatta.com>  Tue, 31 Jan 2012 10:51:42 -0800

vyatta-op-dhcp-server (0.13.18) unstable; urgency=low

  * Return string representation of the BigInt object instead of the
    object its self

 -- John Southworth <john.southworth@vyatta.com>  Wed, 02 Nov 2011 12:13:39 -0500

vyatta-op-dhcp-server (0.13.17) unstable; urgency=low

  * Allow filtering by pool for new dhcp stats

 -- John Southworth <john.southworth@vyatta.com>  Wed, 02 Nov 2011 10:48:35 -0500

vyatta-op-dhcp-server (0.13.16) unstable; urgency=low

  * Check in new 'show dhcp statistics' code, this has been broken for
    sometime and it is needed for the WebGUI

 -- John Southworth <john.southworth@vyatta.com>  Tue, 01 Nov 2011 20:07:55 -0500

vyatta-op-dhcp-server (0.13.15) unstable; urgency=low

  * Commit dhcp server command changes 'clear dhcp' moved to 'reset
    dhcp'

 -- John Southworth <john.southworth@vyatta.com>  Thu, 27 Oct 2011 17:47:47 -0500

vyatta-op-dhcp-server (0.13.14) unstable; urgency=low

  * new branch


src/rl_str_proc.cc
/*
 * Module: rl_str_proc.cc
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the 
 * Free Software Foundation.
 */
#include "rl_str_proc.hh"

using namespace std;

/**
 *
 **/
StrProc::StrProc(const string &in_str, const string &token)
{
  string tmp = in_str;
  
  //convert tabs to spaces
  size_t pos = 0;
  string tabtospace = "    ";
  string::iterator iter = tmp.begin();
  while ((pos = tmp.find("\t", pos)) != string::npos) {
    tmp.replace(pos, 1, tabtospace);
    pos += tabtospace.length();
  }
  
  //remove the cr
  pos = tmp.find("\n");
  if (pos != string::npos) {
    tmp.replace(pos, 1, "");
  }

  //now handle the case of the multiple length token
  //note that we are using the '~' as a token internally
  size_t start = 0, end;
  while ((start = tmp.find(token, start)) != string::npos) {
    tmp.replace(start, token.length(), "~");
  }


  while ((start = tmp.find_first_not_of("~")) != string::npos) {
    tmp = tmp.substr(start, tmp.length() - start);
    end = tmp.find_first_of("~");
    _str_coll.push_back(tmp.substr(0, end));
    tmp = tmp.substr(end+1, tmp.length() - end-1);
    if (end == string::npos) {
      break;
    }
  }
}

/**
 *
 **/
string
StrProc::get(int i)
{
  if (((unsigned int) i) >= _str_coll.size()) {
    return string("");
  }
  return _str_coll[i];
}

/**
 *
 **/
string
StrProc::get(int start, int end)
{
  if (((unsigned int) start) >= _str_coll.size()) {
    return string("");
  }

  string tmp;
  for (int i = start; (i < end) && (((unsigned int) i) < _str_coll.size());
       ++i) {
    tmp += _str_coll[i] + " ";
  }
  return tmp.substr(0,tmp.length()-1);
}

/**
 *
 **/
vector<string>
StrProc::get()
{
  return _str_coll;
}


lib/DHCPServerOpMode.pm
#!/usr/bin/env perl
#
# Module Vyatta::DHCPServerOpMode
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2008 Vyatta, Inc.
# All Rights Reserved.
#
# Portions created by SO3 Group are Copyright (C) 2013 SO3 Group
#
# Author: John Southworth
# Date: January 2011
# Description: Library containing functions for DHCP operational commands
#
# **** End License ****
#

package Vyatta::DHCPServerOpMode;

use lib "/opt/vyatta/share/perl5/";
use strict;
use Math::BigInt;
use File::Slurp;

sub iptoint {
    # Based on the perl Net::IP module
    my $ip = shift;
    my $binip = unpack('B32', pack('C4C4C4C4', split(/\./, $ip)));
    # $n is the increment, $dec is the returned value
    my ($n, $dec) = (Math::BigInt->new(1), Math::BigInt->new(0));
    # Reverse the bit string
    foreach (reverse(split '', $binip)) {
        # If the nth bit is 1, add 2**n to $dec
        $_ and $dec += $n;
        $n *= 2;
    }
    # Strip leading + sign
    $dec =~ s/^\+//;
    
    return $dec->bstr();
}

sub get_active {
    open( my $leases, '<', "/config/dhcpd.leases" );
    my $pool;
    my $ip;
    my $active = 0;
    my %active_hash = ();
    my %active_leases = ();
    while (<$leases>){
        my $line = $_;
        if ($line =~ /lease\s(.*)\s{/){
            $ip = $1;
        }
        next if (!defined($ip));
        if ($line =~ /shared-network:\s(.*)/) {
            $pool = $1;
        }
        next if (!defined($pool));
        if (!defined($active_hash{"$pool"}->{"$ip"})){
            $active_hash{"$pool"}->{"$ip"} = 0;
        } else {
            if ($line =~ /binding state active;/) {
                $active_hash{"$pool"}->{"$ip"} += 1 ;
                ($pool, $ip) = (undef, undef);
            } elsif ($line =~ /binding state free;/ && !($line =~ /next/)) {
                $active_hash{"$pool"}->{"$ip"} -= 1 ;
                ($pool, $ip) = (undef, undef);
            }
        }
    }
    for my $pool (keys %{active_hash}){
        for my $ip ( keys %{$active_hash{$pool}}) {
            if (!defined($active_leases{$pool})){
                $active_leases{$pool} = 0;
            }
            $active_leases{$pool} += 1 if ( $active_hash{"$pool"}->{"$ip"} >= 0 );
        }
    }
    
    return \%active_leases;
}

sub get_pool_size {
    open( my $conf, '<', "/opt/vyatta/etc/dhcpd.conf" )
        or die "Can't open dhcpd.conf";
    my $level = 0;
    my $shared_net;
    my %shared_net_hash = ();
    while (<$conf>) {
        my $line = $_;
        $level++ if ( $line =~ /{/ );
        $level-- if ( $line =~ /}/ );
        if ($line =~ /shared-network\s(.*)\s{/) {
            $shared_net = $1;
        } elsif ($line =~ /range\s(.*?)\s(.*?);/) {
            my $start = iptoint($1);
            my $stop = iptoint($2);
            $shared_net_hash{"$shared_net"} += ($stop - $start + 1);
        } 
    }
    #sanity check the file
    if ($level != 0){
        die "Invalid dhcpd.conf, mismatched braces";
    }
    
    return \%shared_net_hash;
}

sub print_stats {
    my $pool_filter = $_[0];
    my $pool_sizes = get_pool_size();
    my $active = get_active();
    my $format = "%-25s %-11s %-11s %s\n";
    print "\n";
    printf($format, "Pool", "Pool size", "# Leased", "# Avail");
    printf($format, "----", "---------", "--------", "-------");
    for my $pool (keys %{$pool_sizes}) {
        if (defined ($pool_filter)) {
            next if ($pool ne $pool_filter);
        }
        my $pool_size = $pool_sizes->{$pool};
        my $used = $active->{$pool};
        $used = 0 if (!defined($used));
        $pool_size = 0 if (!defined($pool_size));
        my $avail = $pool_size - $used;
        printf($format, $pool, $pool_size, $used, $avail);
    }
    print "\n";
}

# Read leases file into string
#
# If this really causes performance troubles,
# replace with line- or character-oriented parser
sub read_lease_file {
    my $leases_file = read_file('/config/dhcpd.leases');

    return $leases_file;
}

# Extract individual leases and return array
sub get_leases {
    my $leases_raw = shift;
    my @leases = $leases_raw =~ /(lease \s+ [\d\.]+ \s+ {.*?})/gsx;

    return @leases;
}

# Parse individual lease, return parameter hash
sub parse_lease {
    my $lease = shift;
    my %lease_hash = ();

    # Get state
    my ($state) = $lease =~ /binding \s+ state \s+ (.*?) \s* ;/sx;
    die("Malformed lease: missing state!") unless defined($state);
    $lease_hash{"state"} = $state;
    
    # Get client IP address
    my ($ip_address) = $lease =~ /lease \s+ ([\d\.]+) \s+ {/sx;
    die("Malformed lease: missing IP address!") unless defined($ip_address);
    $lease_hash{"ip_address"} = $ip_address;

    # Get hardware address
    # May be absent in backup and free leases so don't complain in that case
    my ($hardware_address) = $lease =~ /hardware \s+ ethernet \s+ (.*?) \s* ;/sx;
    if ($state !~ /backup|free|abandoned/) {    
        die("Malformed lease: missing hardware address!") unless defined($hardware_address);
    } else {
        $hardware_address = "" unless defined($hardware_address);
    }
    $lease_hash{"hardware_address"} = $hardware_address;

    # Get start time
    # May be absent in free leases so don't complain in that case
    my ($start_time) = $lease =~ /starts \s+ \d \s+ (.*?) \s* ;/sx;
    if ($state ne "free") {
        die("Malformed lease: missing start time!") unless defined($start_time);
    } else {
        $start_time = "" unless defined($start_time);
    }
    $lease_hash{"start_time"} = $start_time;

    # Get end time
    # May be absent in backup and free leases so don't complain in that case
    my ($end_time) = $lease =~ /ends \s+ \d \s+ (.*?) \s* ;/sx;
    if ($state !~ /backup|free/) {
        die("Malformed lease: missing end time!") unless defined($end_time);
    } else {
        $end_time = "" unless defined($end_time);
    }
    $lease_hash{"end_time"} = $end_time;

    # Get pool
    my ($pool) = $lease =~ /shared-network: \s+ (.*?) \s+/sx;
    $pool = "" unless defined($pool);
    $lease_hash{"pool"} = $pool;

    # Get client name
    my ($client_name) = $lease =~ /client-hostname \s+ "* (.*?) "* \s* ;/sx;
    if (defined($client_name)) {
        my $prefix = $pool."_";
        $client_name =~ s/$prefix//;
    } else {
        $client_name = "";
    }
    $lease_hash{"hostname"} = $client_name;

    return %lease_hash;
}

sub print_leases {
    my ($pool, $state) = @_;

    # Show active leases by default
    $state = "active" unless defined($state);

    my @leases = get_leases(read_lease_file());

    my $format = "%-16s %-18s %-20s %-25s %s\n";

    printf("\n");
    printf($format, "IP address", "Hardware address", "Lease expiration", "Pool", "Client Name");
    printf($format, "----------", "----------------", "----------------", "----", "-----------");

    for my $lease (@leases) {
        my %lease_hash = parse_lease($lease);

        my $skip = 0;
        if ( (defined($state) && ($lease_hash{"state"} ne $state)) ||
             (defined($pool) && ($lease_hash{"pool"} ne $pool) ) ) {
            $skip = 1;
        }

        if ( !$skip ) {
            printf($format,
                   $lease_hash{"ip_address"},
                   $lease_hash{"hardware_address"},
                   $lease_hash{"end_time"},
                   $lease_hash{"pool"},
                   $lease_hash{"hostname"} );
        }
    }
}

return 1;


scripts/dhcpdv6-show-leases.pl
#!/usr/bin/perl

# Module: dhcpdv6-leases.pl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# A copy of the GNU General Public License is available as
# `/usr/share/common-licenses/GPL' in the Debian GNU/Linux distribution
# or on the World Wide Web at `http://www.gnu.org/copyleft/gpl.html'.
# You can also obtain it by writing to the Free Software Foundation,
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2010 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Bob Gilligan
# Date: April 2010
# Description: Script to display DHCPv6 server leases in a user-friendly form
#
# **** End License ****

use strict;
use lib "/opt/vyatta/share/perl5/";

use Getopt::Long;
use Vyatta::Config;


# Globals
my $lease_filename = "/config/dhcpdv6.leases";
my $debug_flag = 0;

GetOptions(
    "debug"     => \$debug_flag,
    "lease-file=s" =>  \$lease_filename,
    );


sub log_msg {
    my $message = shift;

    print "DEBUG: $message" if $debug_flag;
}


#
# Main section.
#
my @lines=();

if (!open(LEASE_FILE, "<$lease_filename")) {
    printf("DHCPv6 server is not running");
    exit 1;
}

@lines = <LEASE_FILE>;
close(LEASE_FILE);
chomp @lines;

my $level = 0;
my $s1;
my $s2;
my $shared_network;
my $ia_na;
my $iaaddr;
my $ends_day;
my $ends_time;
my $binding_state;
my %ghash = ();

# Parse the leases file into a hash keyed by IPv6 addr.
foreach my $line (@lines) {
    log_msg("Line: $line\n");
    if ($line =~ /^ia-na "(.*)" \{/) {
	if ($level != 0) {
	    printf("Found ia-na at level $level\n");
	    exit 1;
	}
	log_msg("setting ia_na\n");
	$ia_na = $1;
	$level++;
    } elsif ($line =~ /^.*iaaddr .*\{/) {
	if ($level != 1) {
	    printf("Found iaaddr at level $level\n");
	    exit 1;
	}
	($s1, $iaaddr, $s2) = split(' ', $line);
	log_msg("Setting iaaddr to $iaaddr.\n");
	log_msg("s1 $s1 s2 $s2\n");
	$level++;
    } elsif ($line =~ /^.*ends /) {
	if ($level != 2) {
	    printf("Found ends at level $level\n");
	    exit 1;
	}
	log_msg("Setting ends_day ends_time\n");
	($s1, $s2, $ends_day, $ends_time) = split(' ', $line);
	$ends_time =~ s/;//;
    } elsif ($line =~ /^.*binding state /) {
	if ($level != 2) {
	    printf("Found binding state at level $level\n");
	    exit 1;
	}
	log_msg("Setting binding state\n");
	($s1, $s2, $binding_state) = split(' ', $line);
	$binding_state =~ s/;//;
    } elsif ($line =~ /^#shared-network: .*/) {
        log_msg("Setting shared network\n");
        ($s1, $shared_network) = split(' ', $line);
    } elsif ($line =~ /^.*\{/) {
	log_msg("Unknown clause: $line\n");
	$level++;
    } elsif ($line =~ /\}/) {
	$level--;
	if ($level == 0) {
	    if (!defined($ia_na)) {
		printf("ia_na not defined\n");
		exit 1;
	    }

	    if (!defined($iaaddr)) {
		printf("iaaddr not defined\n");
		exit 1;
		}

	    if (!defined($ends_day)) {
		printf("ends_day not defined\n");
		exit 1;
	    }

	    if (!defined($ends_time)) {
		printf("ends_time.\n");
		exit 1;
	    }

	    log_msg("Setting ghash entry for $iaaddr to $ends_day $ends_time\n");
	    my @array = ($ia_na, $ends_day, $ends_time, $binding_state, $shared_network);
	    $ghash{$iaaddr} = \@array;
	    undef $ia_na;
	    undef $iaaddr;
	    undef $ends_day;
	    undef $ends_time;
	    undef $shared_network;
	}
    }
}

# Display the leases...

my $num_entries = scalar(keys %ghash);
if ($num_entries == 0) {
    printf("There are no DHCPv6 leases.\n");
    exit 0;
} elsif ($num_entries == 1) {
    printf("There is one DHCPv6 lease:\n");
} else {
    printf("There are $num_entries DHCPv6 leases:\n");
}

printf("\n");
printf("IPv6 Address                            IAID        DUID                                                        Expiration          State     Pool\n");
printf("--------------------------------------- ----------- ----------------------------------------------------------- ------------------- --------- ----\n");
foreach my $key (keys %ghash) {
    my $entry = $ghash{$key};
    my ($ia_na, $day, $time, $state, $shared_network) = @$entry;
    $ia_na =~ s/\\(\d{3})/"qq|\\$1|"/gee;
    my $iaid = join(':', unpack("H*", substr($ia_na, 0, 4)) =~ m/../g);
    my $duid = join(':', unpack("H*", substr($ia_na, 4)) =~ m/../g);
    printf ("%-39s %s %-59s %s %s %-9s %s\n", $key, $iaid, $duid, $day, $time, $state, $shared_network);
}


scripts/vyatta-show-dhcp-server.pl
#!/usr/bin/env perl
#
# Script: vyatta-show-dhcp-server
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2008 Vyatta, Inc.
# All Rights Reserved.
#
# Author: John Southworth
# Date: January 2011
# Description: Wrapper script for DHCP operational commands 
#
# **** End License ****
#

use lib '/opt/vyatta/share/perl5';
use Vyatta::Config;
use Vyatta::DHCPServerOpMode;
use Getopt::Long;

# Check if DHCP Server is configured
my $cfg = new Vyatta::Config;
if (!$cfg->isEffective("service dhcp-server")) {
    print "DHCP server not configured\n";
    exit 0;
}

my ($show_stats, $show_leases, $pool, $state);
GetOptions("show-stats!" => \$show_stats, 
           "show-leases!" => \$show_leases,
           "state=s" => \$state,
           "pool=s"      => \$pool);


if (defined $show_stats){
  Vyatta::DHCPServerOpMode::print_stats($pool);
}

if (defined($show_leases)) {
    Vyatta::DHCPServerOpMode::print_leases($pool, $state);
}


src/command_proc_show_dhcp_leases.cc
/**
 *    Module: command_proc_show_dhcp_leases.cc
 *
 *    Author: Michael Larson
 *    Date: 2005
 *    Description:
 *
 *    This program is free software; you can redistribute it and/or modify 
 *    it under the terms of the GNU General Public License as published 
 *    by the Free Software Foundation; either version 2 of the License, 
 *    or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be  useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 *    02110-1301 USA
 *
 *    Copyright 2006, Vyatta, Inc.
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>

#include <iostream>
#include <list>
#include <string>
#include <time.h>
#include <stdlib.h>

//#include "config.h"
#include "rl_str_proc.hh"
#include "command_proc_show_dhcp_leases.hh"
#include "xsl_processor.hh"

using namespace std;





int main(int argc, char ** argv) {

	//Build out string request based on the number of argcs.
	string request;
	bool debug = false;
	for (int i = 1; i < argc; ++i) {
		const char *arg = argv[i];
		if (strcmp(arg, "--debug") == 0) {
			debug = true;
		} else {
			request += string(arg) + string(" ");
		}
	}
	if (debug) {
		cout << "request: " << request << endl;
	}


	CommandProcShowDHCPLeases proc;

	// process command and create xml output
	string reason;
	string xml_out = proc.process(request, debug, reason);
	if (debug) {
		cout << "output xml:" << endl << xml_out << endl;
	}

	if (xml_out.empty() == true) {
		cout << reason << endl;
		exit(0);
	}

	bool catch_param_name = false;
	bool catch_param_val = false;
	string param_name;
	string param_val;
	list<pair<string,string> > listParams;
	for (int i = 1; i < argc; ++i) {
		if (strcmp(argv[i], "--pname") == 0) {
			catch_param_name = true;
			catch_param_val = false;
			param_name = "";
			param_val = "";
		} else if (strcmp(argv[i], "--pval") == 0) {
			catch_param_name = false;
			catch_param_val = true;
			param_val = "";
		} else {
			if (catch_param_name) {
				param_name = argv[i];
				catch_param_name = false;
			}
			if (catch_param_val) {
				param_val = argv[i];
				catch_param_val = false;
			}
		}
		if (param_name.length() > 0 && param_val.length() > 0) {
			listParams.push_back(pair<string,string>(param_name, param_val));
			param_name = "";
			param_val = "";
		}
	}





	XSLProcessor xsl_processor(debug);

	cout << xsl_processor.transform(xml_out, proc.xsl(), listParams) << endl;
}




/**
 *
 **/
CommandProcShowDHCPLeases::CommandProcShowDHCPLeases()
{

}

/**
 *
 **/
CommandProcShowDHCPLeases::~CommandProcShowDHCPLeases()
{


}

/**
 *
 **/
std::string
CommandProcShowDHCPLeases::process(const string &cmd, bool debug, string &reason)
{
  bool active_leases;
  string pool;
  StrProc proc_str(cmd, " ");

  if (proc_str.size() == 3) {
    _xsl = XSLDIR "/" + proc_str.get(0);
    pool = proc_str.get(1);
    active_leases = false;
  }
  else if (proc_str.size() == 2 && proc_str.get(1) == "expired") {
    _xsl = XSLDIR "/" + proc_str.get(0);
    active_leases = false;
  }
  else if (proc_str.size() == 2) {
    _xsl = XSLDIR "/" + proc_str.get(0);
    pool = proc_str.get(1);
    active_leases = true;
  }
  else {
    _xsl = XSLDIR "/" + proc_str.get(0);
    active_leases = true;
  }

  reason = "";

  struct stat buf;
  if (stat("/var/run/dhcpd.pid", &buf) != 0) {
    reason = "DHCP server is not running";
    return string("");
  }
  
  //first thing, execute this command;
  if (cmd.empty() == false) {
    //    syslog(LOG_DEBUG | LOG_USER , "********* rl_cmd::execute(): %s\n", cmd.c_str());
    FILE *f = popen("cat /config/dhcpd.leases", "r");
    if (f) {
      char buf[2048];
      
      while(fgets(buf, 2048, f) != NULL) { 
//	printf("process line: %s\n", buf);
	string line(buf);
	convert_to_xml(line, pool, active_leases);
      } 
      if (pclose(f) != 0) {
	return string("");
      }
      complete_xml();
    }
  } 
  else {
    //    syslog(LOG_DEBUG | LOG_USER , "********* rl_cmd::execute(): command is empty\n");
  }

/*  if (_coll.empty() == true) {
    reason = "no clients leased";
    return string("");
  }*/

  _xml_out = "<opcommand name='dhcpleases'><format type='row'>";
  map<string, string>::iterator iter = _coll.begin();
  while (iter != _coll.end()) {
    _xml_out += "<row>" + iter->second + "</row>";
    ++iter;
  }
  _xml_out += "</format></opcommand>";


  string foo = cmd;
  debug = true;
  return _xml_out;
}

/**
 *
 * Sample lease file below
 *
 *

# All times in this file are in UTC (GMT), not your local timezone.   This is
# not a bug, so please don't ask about it.   There is no portable way to
# store leases in the local timezone, so please don't request this as a
# feature.   If this is inconvenient or confusing to you, we sincerely
# apologize.   Seriously, though - don't ask.
# The format of this file is documented in the dhcpd.leases(5) manual page.
# This lease file was written by isc-dhcp-V3.0.3

lease 10.0.0.236 {
  starts 4 2006/02/23 18:34:48;
  ends 5 2006/02/24 18:34:48;
  binding state active;
  next binding state free;
  hardware ethernet 00:12:3f:b3:02:b2;
}

*
*
**/
void 
CommandProcShowDHCPLeases::convert_to_xml(const string &line, const string &pool, bool active_leases)
{
  StrProc str_proc(line, " ");
  static int active_lease = 0;
  static int lease_state_backup = 1;

  if (line.find("lease") != string::npos) {
    _ip = str_proc.get(1);
    _coll.erase(_ip);
    _xml_frag = "<ip>" + str_proc.get(1) + "</ip>";
  }

  if (line.find("starts") != string::npos) {
    _xml_frag += "<start>" + str_proc.get(2) + " " + str_proc.get(3).substr(0, str_proc.get(3).length()-1) + "</start>";
  }

  if (line.find("ends") != string::npos) {
    _xml_frag += "<end>" + str_proc.get(2) + " " + str_proc.get(3).substr(0, str_proc.get(3).length()-1) + "</end>";
  }

  if (line.find("  binding") != string::npos) {
     if (str_proc.get(2) == "active;") {
         active_lease = 1;
         lease_state_backup = 0;
     }
     else if (str_proc.get(2) == "free;") {
         active_lease = 0;
         lease_state_backup = 0;
     }
     else {
 	 lease_state_backup = 1;
     }
        _xml_frag += "<bind_state>" + str_proc.get(2) + " " + "</bind_state>";
  }

  if (line.find("next binding") != string::npos) {
    _xml_frag += "<next_bind_state>" + str_proc.get(3) + " " + "</next_bind_state>";
  }

  if (line.find("hardware") != string::npos) {
    _xml_frag += "<hardware>" + str_proc.get(2).substr(0, str_proc.get(2).length()-1) + "</hardware>";
  }

  if (line.find("uid") != string::npos) {
    //skip uid as it is not used and has illegal xml characters
    //    _xml_frag += "<uid>" + str_proc.get(1).substr(0, str_proc.get(1).length()-1) + "</uid>";
    _xml_frag += "<uid></uid>";
  }
   
  if (line.find("client-hostname") != string::npos) {
    string::size_type indexQS = line.find_first_of("\"");
    string::size_type indexQE = line.find_last_of("\"");
    if (indexQS != string::npos && indexQE != string::npos && indexQS < indexQE) {
      std::string strClientHostname = line.substr(indexQS + 1, indexQE - indexQS - 1);
      if (strClientHostname.length() > 0) {
        _xml_frag += "<client_hostname>" + strClientHostname + "</client_hostname>";
      }
    }
  }

  if (line.find("shared-network") != string::npos) {
    _pool = str_proc.get(1).substr(0, str_proc.get(1).length());
    _xml_frag += "<shared_network>" + str_proc.get(1).substr(0, str_proc.get(1).length()) + "</shared_network>";
  }


  if (line.find("}") != string::npos) {
    //printf("before insertion: %s\n", _xml_frag.c_str());
    if (pool.empty() == false) {
      if (pool == _pool) {
       if (lease_state_backup == 0) {
         if (active_leases == true){ 
             if (active_lease == 1) {             
	     _coll.insert(pair<string, string>(_ip, _xml_frag));
            }
	 } else if (active_leases == false) {
                   if (active_lease == 0) {
 		      _coll.insert(pair<string, string>(_ip, _xml_frag));
		   }	
         } 	      
       } 
     }
    }
    else {
          if (lease_state_backup == 0) {
           if (active_leases == true){
             if (active_lease == 1) {
                 _coll.insert(pair<string, string>(_ip, _xml_frag));
	     }
         } else if (active_leases == false) {
		   if (active_lease == 0) {
	   		_coll.insert(pair<string, string>(_ip, _xml_frag));
		   } 
	}
      }
    } 
  } 
  return;
}

/**
 *
 **/
void
CommandProcShowDHCPLeases::complete_xml()
{
  if (_xml_out.empty() == false) {
    _xml_out += "</format></opcommand>";
  }
}
